keps/sig-autoscaling/4943-vpa-heterogeneous-workloads/README.md

```diff
@@ -0,0 +1,911 @@
+notes and implementing the feature itself. KEP editors and SIG Docs
+should help to ensure that the tone and content of the `Summary` section is
+useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+
+Both in this section and below, follow the guidelines of the [documentation
+style guide]. In particular, wrap lines to a reasonable length, to make it
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+This KEP proposes extending the Vertical Pod Autoscaler (VPA) to support use
+cases where each pod under a DaemonSet should have different resource requests
+and limits. This is implemented by adding a `Scope` field to the
+VerticalPodAutoscaler API, which initially supports a value of `node`,
+indicating that the VPA recommendation should be calculated and applied
+independently for pods scheuled to the same node.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports -->
+
+Vertical Pod Autoscaling allows users to automatically adjust the resource
+requests and limits for pods managed by a controller. The pre-existing behavior
+applies the same resource requests and limits to every pod controlled under the
+target reference. In cases where the workloads, by design, do not evenly
+distribute load, uniform resource allocation is undesirable.
+
+DaemonSets that function as node-local agents are a common example. One
+[experience report for a Managed Prometheus
+service](https://blog.bernot.io/p/vpa-for-workloads-with-heterogeneous-resource-requirements/)
+describes this problem in more detail. There are similar examples in Networking
+and Observability where the load for any individual pod under a DaemonSet will
+depend on activity on the node that is not necessarily proportional to any
+other factors.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+- Vertically scale pods under a DaemonSet with different resource requests and
+  limits per node.
+- Effectively manage large disparities in resource requirements between pods in
+  a DaemonSet.
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+- Consider [node
+  metatdata](https://github.com/kubernetes/autoscaler/issues/5928).
+- Change autoscaling behavior for pods that are not run under a DaemonSet.
+- Provide an informed recommendation when a new node is added to a cluster.
```

Comment:
Since providing an informed recommendation is out of scope, what recommendation should be applied to recommendations on new nodes?

Reply:
Maybe we should reword this to something like "provide a different recommendation than what would be provided by VPA normally when a new node is added"